src/python/aqueduct_executor/operators/utils/storage/s3.py

```diff
@@ -63,14 +66,33 @@ def exists(self, key: str) -> bool:
                 return False
         return True
 
-    def _prefix_key(self, key: str) -> str:
+    def _resolve_full_key(self, key: str) -> str:
+        """The `key_prefix` is expected to be in the format `path/to/dir/`."""
         if not self._key_prefix:
             return key
-        return self._key_prefix + "/" + key
 
+        assert self._key_prefix[0] != "/" and self._key_prefix[-1] == "/"
+        return self._key_prefix + key
+
+
+def _sanitize_path(path: str) -> str:
+    """Sanitize the given path to be in the format `path/to/dir/` (no leading slash but one trailing slash)."""
+    if path == "":
+        return path
+    if path[0] == "/":
+        path = path[1:]
+    if path[-1] != "/":
+        path += "/"
+    return path
 
-def parse_s3_path(s3_path: str) -> Tuple[str, str]:
-    path_parts = s3_path.replace("s3://", "").split("/")
+
+def parse_s3_bucket_and_key_prefix(s3_bucket_path: str, root_dir_path: str) -> Tuple[str, str]:
+    """The root_dir_path is assumed to be in the format `path/to/dir/`."""
+    path_parts = s3_bucket_path.replace("s3://", "").split("/")
     bucket = path_parts.pop(0)
-    key = "/".join(path_parts)
-    return bucket, key
+
+    if root_dir_path != "":
+        path_parts += [_sanitize_path(root_dir_path)]
```

Comment:
if root_dir_path is assumed to be in the format `path/to/dir/`, why do we need to sanitize it again?

Reply:
It is assumed, but we don't check it here. I didn't want to be too strict with the assertions in case there was a codepath I missed. Agree, it's confusing - maybe I can just remove the comment for now and make no assumptions about the root path.

Reply:
reminder to remove the comment.